core/primitives-core/src/profile.rs

```diff
@@ -19,154 +21,394 @@ impl Default for ProfileData {
+            Cost::ExtCost(ExtCosts::storage_remove_key_byte) => &mut self.costs[44],
+            Cost::ExtCost(ExtCosts::storage_remove_ret_value_byte) => &mut self.costs[45],
+            Cost::ExtCost(ExtCosts::storage_has_key_base) => &mut self.costs[46],
+            Cost::ExtCost(ExtCosts::storage_has_key_byte) => &mut self.costs[47],
+            Cost::ExtCost(ExtCosts::storage_iter_create_prefix_base) => &mut self.costs[48],
+            Cost::ExtCost(ExtCosts::storage_iter_create_prefix_byte) => &mut self.costs[49],
+            Cost::ExtCost(ExtCosts::storage_iter_create_range_base) => &mut self.costs[50],
+            Cost::ExtCost(ExtCosts::storage_iter_create_from_byte) => &mut self.costs[51],
+            Cost::ExtCost(ExtCosts::storage_iter_create_to_byte) => &mut self.costs[52],
+            Cost::ExtCost(ExtCosts::storage_iter_next_base) => &mut self.costs[53],
+            Cost::ExtCost(ExtCosts::storage_iter_next_key_byte) => &mut self.costs[54],
+            Cost::ExtCost(ExtCosts::storage_iter_next_value_byte) => &mut self.costs[55],
+            Cost::ExtCost(ExtCosts::touching_trie_node) => &mut self.costs[56],
+            Cost::ExtCost(ExtCosts::promise_and_base) => &mut self.costs[57],
+            Cost::ExtCost(ExtCosts::promise_and_per_promise) => &mut self.costs[58],
+            Cost::ExtCost(ExtCosts::promise_return) => &mut self.costs[59],
+            Cost::ExtCost(ExtCosts::validator_stake_base) => &mut self.costs[60],
+            Cost::ExtCost(ExtCosts::validator_total_stake_base) => &mut self.costs[61],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_g1_multiexp_base) => &mut self.costs[62],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_g1_multiexp_byte) => &mut self.costs[63],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_g1_multiexp_sublinear) => &mut self.costs[64],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_pairing_check_base) => &mut self.costs[65],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_pairing_check_byte) => &mut self.costs[66],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_g1_sum_base) => &mut self.costs[67],
+            #[cfg(feature = "protocol_feature_alt_bn128")]
+            Cost::ExtCost(ExtCosts::alt_bn128_g1_sum_byte) => &mut self.costs[68],
+            Cost::ExtCost(ExtCosts::__count) => unreachable!(),
+        }
+    }
+}
+
+impl ProfileData {
+    pub fn nonzero_costs(&self) -> Vec<(Cost, u64)> {
+        let mut data = Vec::new();
+        for i in Cost::ALL {
+            if self[*i] > 0 {
+                data.push((*i, self[*i]));
+            }
+        }
+        data
+    }
+}
```

Comment:
This looks unfortunate to me... How often do we do these operations? I would estimate that these indexes would introduce some runtime and code size penalty due to the fact that compiler will add boundary checks everywhere (since it is not Vec of any length rather than an array of a constant length), another unnecessary panic source is `unreachable!` which is there only to save us some compilation time (really?) since we introduced `__count` instead of using some macro which would just generate a const for us.

Reply:
> another unnecessary panic source is unreachable! which is there only to save us some compilation time (really?) since we introduced __count instead of using some macro which would just generate a const for us.

I don't understand this statement, why `__count => unreachable!` is unnecessary? Without it match cases isn't complete.

Reply:
That part of my comment was more about the pattern we decided to use to count the number of enum variants by this workaround with `__count`. I recall I had a conversation with someone (@matklad was it you?) about using this workaround in the first place, and the alternatives we discussed: (1) keep the number of variants as a variable that has to be manually updated (prone to errors, especially on merge conflicts), (2) use macro (potential increase in compilation time was the argument against it). In this context I just highlighted that this current workaround injects runtime costs (panic handler) in, what I assume, a hot execution path

Reply:
Ah I see, so even though `__count=>unreachable!` is really unreachable in execution of the node, have it here make rustc cannot do best optimization to the whole function, which you mentioned  as" injects runtime costs (panic handler)" ?

Reply:
>In this context I just highlighted that this current workaround injects runtime costs (panic handler) in, what I assume, a hot execution path

[At the hot call-sites](https://github.com/near/nearcore/blob/96afe0329006a1dd3b4be065c687407268bbfb24/runtime/near-vm-logic/src/logic.rs#L2300), the costs are compile-time constants,  so the compiler should be able to constant-fold both the match *and* the bounds check on the actual array access. 

In a separate PR, I want to merge both costs enums into one, so that all explicit matching goes away, we can write just `data[cost as usize]`, and adding a new cost (of either kind) can be done in backwards-compatible fashion by appending to the end.